### PR TITLE
chore: merge the dual state into the platform state

### LINKED ISSUE #10414: Merge Platform State and Dual State
Execute the changes discussed in design sessions by merging the platform state and the dual state.

### PR DESCRIPTION
**Pull request by cody-littley**
Closes #10414

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 1]
```diff
@@ -83,56 +230,290 @@ public PlatformState copy() {
...
+     * Set the consensus timestamp for this state, defined as the timestamp of the first transaction that was applied in
+     * the round that created the state.
+     *
+     * @param consensusTimestamp a consensus timestamp
+     */
+    public void setConsensusTimestamp(final Instant consensusTimestamp) {
+        this.consensusTimestamp = consensusTimestamp;
+    }
+
+    // TODO remove methods that are not simple getters and setters
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:341]: TODO left in code.

**Review by edward-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 2]
```diff
@@ -16,46 +16,155 @@
...
 
 /**
- * This subtree contains state data which is managed and used exclusively by the platform.
+ * State managed and used by the platform.
  */
-public class PlatformState extends PartialNaryMerkleInternal implements MerkleInternal {
+public class PlatformState extends PartialMerkleLeaf implements MerkleLeaf {
 
-    public static final long CLASS_ID = 0x483ae5404ad0d0bfL;
+    // TODO refactor platform code to not directly call into this class
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:42]: TODOs

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 3]
```diff
@@ -16,46 +16,155 @@
...
+    /**
+     * The round of this state. This state represents the handling of all transactions that have reached consensus in
+     * all previous rounds. All transactions from this round will eventually be applied to this state. The first state
+     * (genesis state) has a round of 0 because the first round is defined as round 1, and the genesis state is before
+     * any transactions are handled.
+     */
+    private long round = GENESIS_ROUND;
+
+    /**
+     * running hash of the hashes of all consensus events have there been throughout all of history, up through the
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:75]: grammar "...events have there been..."

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 4]
```diff
@@ -16,46 +16,155 @@
...
+
+    /**
+     * the consensus timestamp for this signed state
+     */
+    private Instant consensusTimestamp;
+
+    /**
+     * the minimum generation of famous witnesses per round
+     */
+    //    private List<MinGenInfo> minGenInfo; // TOD make sure it's ok to remove this
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:88]: TOD

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 5]
```diff
@@ -16,46 +16,155 @@
...
+     */
+    private ConsensusSnapshot snapshot;
+
+    /**
+     * the time when the freeze starts
+     */
+    private Instant freezeTime;
+
+    /**
+     * the last freezeTime based on which the nodes were frozen
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:122]: grammar

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 6]
```diff
@@ -16,46 +16,155 @@
...
     public PlatformState() {}
 
     /**
      * Copy constructor.
      *
-     * @param that
-     * 		the node to copy
+     * @param that the object to copy
      */
     private PlatformState(final PlatformState that) {
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:138]: Optional: add nullity annotations throughout this class

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 7]
```diff
@@ -83,56 +230,290 @@ public PlatformState copy() {
...
+        this.round = round;
     }
 
     /**
-     * Set the previous address book.
+     * Get the running hash of all events that have been applied to this state since the beginning of time.
+     *
+     * @return a running hash of events
+     */
+    public Hash getHashEventsCons() {
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:308]: Is now a good time to change method names?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 8]
```diff
@@ -83,56 +230,290 @@ public PlatformState copy() {
...
+    /**
+     * The minimum generation of famous witnesses for the round specified. This method only looks at non-ancient rounds
+     * contained within this state.
+     *
+     * @param round the round whose minimum generation will be returned
+     * @return the minimum generation for the round specified
+     * @throws NoSuchElementException if the generation information for this round is not contained withing this state
+     */
+    public long getMinGen(final long round) {
+        for (final MinGenInfo info : getMinGenInfo()) {
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:362]: `getMinGenInfo` might return `null`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 9]
```diff
@@ -83,56 +230,290 @@ public PlatformState copy() {
...
+        throw new NoSuchElementException("No minimum generation found for round: " + round);
+    }
+
+    /**
+     * Return the round generation of the oldest round in this state
+     *
+     * @return the generation of the oldest round
+     */
+    public long getMinRoundGeneration() {
+        return getMinGenInfo().stream()
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:376]: `getMinGenInfo` might return `null`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java [hunk 10]
```diff
@@ -83,56 +230,290 @@ public PlatformState copy() {
...
+    public Instant getLastFrozenTime() {
+        return lastFrozenTime;
+    }
+
+    /**
+     * Sets the last freezeTime based on which the nodes were frozen.
+     *
+     * @param lastFrozenTime the last freezeTime based on which the nodes were frozen
+     */
+    @Nullable
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:496]: `Nullable` annotation on `void` method

**Review by alittley — APPROVED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java [hunk 11]
```diff
@@ -118,7 +118,7 @@ public interface HederaInjectionComponent {
 
     ThrottleManager throttleManager();
 
-    DualStateUpdateFacility dualStateUpdateFacility();
+    PlatformStateUpdateFacility dualStateUpdateFacility();
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/HederaInjectionComponent.java:121]: renamed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 1 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:341]: These TODOs should be done in future PRs, not this one. Removed them for now.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/network/NetworkLogicModule.java [hunk 12]
```diff
@@ -41,8 +41,8 @@ public static UpgradeActions.UnzipAction provideUnzipAction() {
 
     @Provides
     @Singleton
-    public static Supplier<SwirldDualState> provideDualState(DualStateAccessor dualStateAccessor) {
-        return dualStateAccessor::getDualState;
+    public static Supplier<PlatformState> provideDualState(PlatformStateAccessor platformStateAccessor) {
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/network/NetworkLogicModule.java:44]: fixed

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/network/UpgradeActions.java [hunk 13]
```diff
@@ -183,8 +183,8 @@ private void catchUpOnMissedUpgradePrep() {
         extractSoftwareUpgrade(archiveData).join();
     }
 
-    private void withNonNullDualState(final String actionDesc, final Consumer<SwirldDualState> action) {
-        final var curDualState = dualState.get();
+    private void withNonNullDualState(final String actionDesc, final Consumer<PlatformState> action) {
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/network/UpgradeActions.java:186]: fixed

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/network/UpgradeActions.java [hunk 14]
```diff
@@ -183,8 +183,8 @@ private void catchUpOnMissedUpgradePrep() {
...
+        final var curDualState = platformState.get();
         Objects.requireNonNull(curDualState, "Cannot " + actionDesc + " without access to the dual state");
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/network/UpgradeActions.java:188]: fixed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 2 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:42]: removed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 3 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:75]: fixed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 4 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:88]: removed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 5 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:122]: fixed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 6 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:138]: Added. It's annoying, basically everything has to be `@Nullable` given the funky lifecycle of serialized objects.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 7 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:308]: Yes, changed to `getRunningEventHash()`.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 8 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:362]: Pre-existing issue that was copy-pasted into this class. I added the following check, at least the exception will be more meaningful if we ever encounter this in the wild.


```
        final List<MinGenInfo> minGenInfo = getMinGenInfo();
        if (minGenInfo == null) {
            throw new IllegalStateException("No MinGen info found in state for round " + round);
        }

        for (final MinGenInfo info : minGenInfo) {
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 9 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:376]: Pre-existing, added check and specific exception.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java (hunk 10 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/PlatformState.java:496]: fixed

**Review by edward-swirldslabs — APPROVED** [review]

**Review by tinker-michaelj — APPROVED** [review]
LGTM for Service changes, thx @cody-littley !

**Pull request merged**